plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts

```diff
@@ -0,0 +1,487 @@
+                    fetchTimeInterval: nextFetchTimeInterval(payload, events.length),
+                } as ExportHistoricalEventsJobPayload)
+                .runIn(1, 'seconds')
+        }
+        // :TODO: When unknown exportEventsError?
+
+        if (events.length > 0 && !exportEventsError) {
+            createLog(
+                `Successfully processed events ${payload.offset}-${payload.offset + events.length} from ${new Date(
+                    payload.timestampCursor
+                ).toISOString()} to ${new Date(payload.timestampCursor + payload.fetchTimeInterval).toISOString()}.`
+            )
+        }
+    }
+
+    async function getTimestampBoundaries(payload: ExportHistoricalEventsUIPayload): Promise<TimestampBoundaries> {
+        if (payload && payload.dateFrom && payload.dateTo) {
+            const min = new Date(payload.dateFrom)
+            const max = new Date(payload.dateTo)
+
+            if (isNaN(min.getTime()) || isNaN(max.getTime())) {
+                createLog(`'dateFrom' and 'dateTo' should be timestamps in ISO string format.`)
+                throw new Error(`'dateFrom' and 'dateTo' should be timestamps in ISO string format.`)
+            }
+            return { min, max }
+        } else {
+            const timestampBoundaries = await fetchTimestampBoundariesForTeam(hub.db, pluginConfig.team_id)
+
+            // no timestamp override specified via the payload, default to the first event ever ingested
+            if (!timestampBoundaries) {
+                throw new Error(
+                    `Unable to determine the timestamp bound for the export automatically. Please specify 'dateFrom'/'dateTo' values.`
+                )
+            }
+
+            return timestampBoundaries
+        }
+    }
+
+    function nextFetchTimeInterval(payload: ExportHistoricalEventsJobPayload, eventCount: number): number {
+        if (eventCount === EVENTS_PER_RUN) {
+            return payload.fetchTimeInterval
+        }
+        // If we're fetching too small of a window at a time, increase window to fetch
+        if (payload.offset === 0 && eventCount < EVENTS_PER_RUN * 0.5) {
+            return Math.min(Math.floor(payload.fetchTimeInterval * 1.2), TWELVE_HOURS)
+        }
+        // If time window seems too large, reduce it
+        if (payload.offset > 2 * EVENTS_PER_RUN) {
+            return Math.max(Math.floor(payload.fetchTimeInterval / 1.2), TEN_MINUTES)
+        }
+        return payload.fetchTimeInterval
+    }
+
+    async function coordinateHistoricExport(update?: CoordinationUpdate) {
+        const params = await getExportParameters()
+
+        if (!params) {
+            // No export running!
+            return
+        }
+
+        const { done, running } = (await meta.storage.get(`EXPORT_COORDINATION`, {})) as CoordinationPayload
+        update = update || (await calculateCoordination(params, done || [], running || []))
```

Comment:
So currently on a restart we'll just re-do the entire date? Any plans to change this or are we just thinking that processing one day's worth of a large customer's events will be quite fast?

Else we'll do a _lot_ of double processing on Cloud

Reply:
No resume logic currently exists (documented under a TODO later here). The original historical export resume logic was flawed since it re-exported everything in a time window (even if way more than 500 events).